### ISSUE TITLE
weird MPS behaviour

### ISSUE THREAD

**Issue by jacksalici** [jacksalici/PytorchTemplate#1]
When using the CPU as the device, the toy problem reaches ~100% accuracy in a few epochs.

```bash
❯ python run.py --config default.yaml toySorter.yaml
{
    "seed": 42,
    "force_reproducibility": true,
    "device": "cpu",
    "lr": 0.0005,
    "batch_size": 64,
    "num_epochs": 5,
    "model_name": "transformer",
    "input_dim": 10,
    "embed_dim": 32,
    "seq_len": 6,
    "num_layers": 3,
    "hidden_dims": null,
    "dropout_rate": 0.1,
    "activation": "relu",
    "use_batch_norm": true,
    "use_layer_norm": false,
    "output_dim": 2,
    "num_heads": 1,
    "attention_dropout": 0.1,
    "transformer_num_layers": 2,
    "experiment": "toysort",
    "dataloader": "toysort_ds",
    "optim": "adamw",
    "avoid_wandb": true,
    "save_model": true,
    "checkpoint_path": "checkpoints",
    "checkpoint_name": null,
    "config": [
        "default.yaml",
        "toySorter.yaml"
    ],
    "config_path": "configs",
    "run_name": "PytorchTemplate_2025-08-20_23-32-00",
    "task": "training"
}
Model: transformer, Parameters: 177,347
Epoch: 0 | Train Loss: 0.8011381821706891 | Test Loss: 0 | Accuracy: 0.275 | Correct Sequences: 176 | Total Sequences: 640 | 
Saving model at epoch 0 with Test Loss 0
Epoch: 1 | Train Loss: 0.3492032347712666 | Test Loss: 0 | Accuracy: 0.9125 | Correct Sequences: 584 | Total Sequences: 640 | 
Saving model at epoch 1 with Test Loss 0
Epoch: 2 | Train Loss: 0.21086752181872725 | Test Loss: 0 | Accuracy: 0.9703125 | Correct Sequences: 621 | Total Sequences: 640 | 
Saving model at epoch 2 with Test Loss 0
Epoch: 3 | Train Loss: 0.16519723006058484 | Test Loss: 0 | Accuracy: 0.9953125 | Correct Sequences: 637 | Total Sequences: 640 | 
Saving model at epoch 3 with Test Loss 0
Epoch: 4 | Train Loss: 0.14352412067819387 | Test Loss: 0 | Accuracy: 0.9921875 | Correct Sequences: 635 | Total Sequences: 640 | 
Saving model at epoch 4 with Test Loss 0
```

Keeping all the same, but using MPS on a 2020 MBA M1, the performance is the worst. 

```bash
❯ /Users/jacksalici/Dev/PytorchTemplate/.venv/bin/python run.py --config default.yaml toySorter.yaml --device mps
{
    "seed": 42,
    "force_reproducibility": true,
    "device": "mps",
    "lr": 0.0005,
    "batch_size": 64,
    "num_epochs": 5,
    "model_name": "transformer",
    "input_dim": 10,
    "embed_dim": 32,
    "seq_len": 6,
    "num_layers": 3,
    "hidden_dims": null,
    "dropout_rate": 0.1,
    "activation": "relu",
    "use_batch_norm": true,
    "use_layer_norm": false,
    "output_dim": 2,
    "num_heads": 1,
    "attention_dropout": 0.1,
    "transformer_num_layers": 2,
    "experiment": "toysort",
    "dataloader": "toysort_ds",
    "optim": "adamw",
    "avoid_wandb": true,
    "save_model": true,
    "checkpoint_path": "checkpoints",
    "checkpoint_name": null,
    "config": [
        "default.yaml",
        "toySorter.yaml"
    ],
    "config_path": "configs",
    "run_name": "PytorchTemplate_2025-08-20_23-35-16",
    "task": "training"
}
Model: transformer, Parameters: 177,347
Epoch: 0 | Train Loss: nan | Test Loss: 0 | Accuracy: 0.003125 | Correct Sequences: 2 | Total Sequences: 640 | 
Saving model at epoch 0 with Test Loss 0
Epoch: 1 | Train Loss: nan | Test Loss: 0 | Accuracy: 0.003125 | Correct Sequences: 2 | Total Sequences: 640 | 
Saving model at epoch 1 with Test Loss 0
Epoch: 2 | Train Loss: nan | Test Loss: 0 | Accuracy: 0.003125 | Correct Sequences: 2 | Total Sequences: 640 | 
Saving model at epoch 2 with Test Loss 0
Epoch: 3 | Train Loss: nan | Test Loss: 0 | Accuracy: 0.003125 | Correct Sequences: 2 | Total Sequences: 640 | 
Saving model at epoch 3 with Test Loss 0
Epoch: 4 | Train Loss: nan | Test Loss: 0 | Accuracy: 0.003125 | Correct Sequences: 2 | Total Sequences: 640 | 
Saving model at epoch 4 with Test Loss 0
```

Still investigating why.
